ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifier.java

```diff
@@ -90,8 +94,68 @@ public void onAttestationsDue(final UInt64 slot) {
     eventThread.execute(() -> internalAttestationsDue(slot));
   }
 
-  public SafeFuture<Optional<Bytes8>> getPayloadId() {
-    return eventThread.execute(() -> lastPayloadId);
+  public SafeFuture<Optional<Bytes8>> getPayloadId(Bytes32 beaconBlockRoot, UInt64 slot) {
+    return eventThread.execute(() -> internalGetPayloadId(beaconBlockRoot, slot));
+  }
+
+  public void onTTDReached(Bytes32 executionBlockHash) {
+    eventThread.execute(() -> internalTTDReached(executionBlockHash));
+  }
+
+  private void internalTTDReached(Bytes32 executionBlockHash) {
+    eventThread.checkOnEventThread();
+    executionPayloadTerminalBlockHash = Optional.of(executionBlockHash);
+  }
+
+  /**
+   * must return Optional.empty() only when is safe to produce a block with an empty execution
+   * payload Optional.of(payloadId) when it is certain that belongs to the given beacon block root
+   * and slot
+   *
+   * <p>in all other cases it must Throw to avoid block production
+   */
+  private Optional<Bytes8> internalGetPayloadId(Bytes32 beaconBlockRoot, UInt64 slot) {
+    eventThread.checkOnEventThread();
+
+    Bytes32 executionBlockHash =
+        recentChainData
+            .getExecutionBlockHashForBlockRoot(beaconBlockRoot)
+            .orElseThrow(
+                () ->
+                    new IllegalStateException(
+                        "Failed to retrieve execution payload hash from beacon block root"));
+
+    if (lastPayloadId.isEmpty()) {
+      if (executionBlockHash.isZero()) {
+        if (executionPayloadTerminalBlockHash.isPresent()) {
+          throw new IllegalStateException(
+              "Execution Block Hash is Zero but we reached a TTD block. Trying to produce a block while not in sync?");
+        }
+
+        LOG.trace("TTD not reached at Beacon Block Root {} and Slot {}", beaconBlockRoot, slot);
+
+        return Optional.empty();
+      }
+
+      throw new IllegalStateException(
+          String.format(
+              "PayloadId not available for Beacon Block Root %s and Slot %s",
+              beaconBlockRoot, slot));
+    }
+
+    if (lastSentForkChoiceState
+            .map(lsfcState -> lsfcState.getHeadBlockHash().equals(executionBlockHash))
+            .orElse(false)
+        && lastPayloadIdSlot.map(fcsSlot -> fcsSlot.compareTo(slot) == 0).orElse(false)) {
+      return lastPayloadId;
+    }
+
+    // TODO: try to obtain a payloadId on the fly?
+
+    throw new IllegalStateException(
+        String.format(
+            "Cannot provide payloadId %s. It doesn't belong to requested Beacon Block Root %s and Slot %s",
+            lastPayloadId.get(), beaconBlockRoot, slot));
```

Comment:
@ajsutton what about this new version?

Reply:
//TODO try to obtaing payloadId should apply also to line 139

Reply:
I think this is pretty close but we can probably simplify it to make things clearer. I think we have four cases:

1. If we have a payloadId it matches the execution block hash, return it.
2. If parentExecutionHash is zero and lastSentForkChoiceHead equals executionPayloadTerminalBlockHash, return the payload ID (this is the merge block)
3. If executionBlockHash is zero, return empty (merge block hasn't been created yet)
4. Otherwise throw an exception

I don't think the slot matters - we can use a payloadId for any slot so long as the parent execution hash we generated the payload ID with matches the execution hash from the parent block.  ie if we have an empty slot we can still use the payload ID from the prior slot because there's no change to the execution layer head.

So I wind up with this method being:

```
  private Optional<Bytes8> internalGetPayloadId(Bytes32 parentBeaconBlockRoot) {
    eventThread.checkOnEventThread();
    final Bytes32 parentExecutionHash =
        recentChainData
            .getExecutionBlockHashForBlockRoot(parentBeaconBlockRoot)
            .orElseThrow(
                () ->
                    new IllegalStateException(
                        "Failed to retrieve execution payload hash from beacon block root"));
    final Bytes32 lastSentForkChoiceHead =
        lastSentForkChoiceState.map(ForkChoiceState::getHeadBlockHash).orElse(Bytes32.ZERO);

    if (lastSentForkChoiceHead.equals(parentExecutionHash)) {
      // Payload ID builds on the correct execution block
      return lastPayloadId;
    }
    if (parentExecutionHash.isZero()) {
      // Haven't reached the merge block yet, parent has default payload
      if (executionPayloadTerminalBlockHash.isPresent()
          && lastSentForkChoiceHead.equals(executionPayloadTerminalBlockHash.get())) {
        // Creating the merge block - payload ID is building on the terminal PoW block
        return lastPayloadId;
      }
      // Can still use a default payload
      return Optional.empty();
    } else {
      // Merge is complete so we must have a real payload, but we don't have one that matches
      // TODO: try to obtain a payloadId on the fly?
      throw new IllegalStateException(
          "Payload ID required but not available for parent block root " + parentBeaconBlockRoot);
    }
  }
```

With that the tests are passing except for the one that expects an exception if the slot doesn't match and the one expecting an exception if the terminal block is found, but those are the two cases I think we need to change, so hopefully I've managed to get my logic right. :)